[Skia] Fix image-based borders

Scony opened the pull request:
#### 619a9da57af1745fc3f312738c9686473b5ca7b2
<pre>
[Skia] Fix image-based borders
<a href="https://bugs.webkit.org/show_bug.cgi?id=271755">https://bugs.webkit.org/show_bug.cgi?id=271755</a>

Reviewed by Carlos Garcia Campos.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawPattern):

Canonical link: <a href="https://commits.webkit.org/277373@main">https://commits.webkit.org/277373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1805d3039551cffceb93dbf587fe7aaf98409484

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26624 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50124 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43489 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49748 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24083 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48022 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/24197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/19944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/21634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5484 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52001 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22475 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23747 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24537 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6681 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23465 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -881,10 +881,13 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
-    phaseMatrix.setTranslate(rect.x(), rect.y());
+    // Adjust phase based on tile offset and spacing.
+    double phaseOffsetX = phase.x() + tileRect.x() * patternTransform.a();
+    double phaseOffsetY = phase.y() + tileRect.y() * patternTransform.d();
```
If understand the problem correctly, we are just missing the scale in the rect. So, just adding `rect.scale(patternTransform.a(), patternTransform.d());` before the moveBy would fix the problem, no? If that's the case, we can probably simplify this by using a FloatPoint instead of a rect, since now we only use the position, but not the size, something like this:

```
FloatPoint phaseOffset(tileRect.location());
phaseOffset.scale(patternTransform.a(), patternTransform.d());
phaseOffset.moveBy(phase);
SkMatrix phaseMatrix;
phaseMatrix.setTranslate(phaseOffset.x(), phaseOffset.y());
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -881,10 +881,13 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
+    double phaseOffsetX = phase.x() + tileRect.x() * patternTransform.a();
+    double phaseOffsetY = phase.y() + tileRect.y() * patternTransform.d();
+    phaseOffsetX -= std::trunc(phaseOffsetX / (tileRect.width() * patternTransform.a() + spacing.width())) * (tileRect.width() * patternTransform.a() + spacing.width());
+    phaseOffsetY -= std::trunc(phaseOffsetY / (tileRect.height() * patternTransform.d() + spacing.height())) * (tileRect.height() * patternTransform.d() + spacing.height());
```
I don't think we need this, it was done in Cairo code due to the surface size limitation, but that's cairo specific.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -881,10 +881,13 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
+    phaseMatrix.setTranslate(phaseOffsetX, phaseOffsetY);
     SkMatrix shaderMatrix = SkMatrix::Concat(phaseMatrix, patternTransform);
     auto samplingOptions = toSkSamplingOptions(m_state.imageInterpolationQuality());
     bool needsClip = tileRect.size() != nativeImage.size();
```
If clip is never really needed this variable name is confusing. Since it will be used only once, we don't need it all

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -895,12 +898,9 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
     if (spacing.isZero() && !needsClip)
```
Here we could just do `if (spacing.isZero() && tileRect.size() == nativeImage.size())`

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -895,12 +898,9 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
-        }
         SkPictureRecorder recorder;
-        auto* recordCanvas = recorder.beginRecording(SkRect::MakeWH(tileRect.width() + spacing.width() / patternTransform.a(), tileRect.height() + spacing.height() / patternTransform.d()));
+        auto* recordCanvas = recorder.beginRecording(tileRect.width() + spacing.width() / patternTransform.a(), tileRect.height() + spacing.height() / patternTransform.d());
```
What is this changing?

Scony replied:
that's some leftover refactoring I tried - it's worth to cleanup such cases in separate PR I guess

Scony committed:
[Skia] Fix image-based borders
https://bugs.webkit.org/show_bug.cgi?id=271755

Reviewed by Carlos Garcia Campos.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawPattern):

Canonical link: https://commits.webkit.org/277373@main
